File: PostAnalyzer/src/ProduceCombineHistograms.cc
#include <TString.h>
#include <iostream>
#include <TStyle.h>
#include <TFile.h>
#include <TH1.h>
#include <TH1D.h>
#include <TCanvas.h>
#include <TText.h>
#include <TPaveText.h>
#include <TGaxis.h>
#include <TGraph.h>
#include <TStyle.h>
#include <TGraphAsymmErrors.h>
#include <TLegend.h>
#include <TLegendEntry.h>
#include <TROOT.h>
#include <TKey.h>
#include <TLatex.h>
#include <TClass.h>
#include <fstream>

#include "Analyzer/LQDM/PostAnalyzer/include/PostAnalysisTool.h"

using namespace std;

void PostAnalysisTool::ProduceCombineHistograms(){


  vector<TString> samples = {"LQLQToBTau_MLQ1000_MPS117_MC1100_L1p0", "LQLQToBTau_MLQ1360_MPS117_MC1100_L1p0", "LQLQToBTau_MLQ1810_MPS117_MC1100_L1p0", "LQLQToBTau_MLQ2170_MPS117_MC1100_L1p0", "LQLQToBTau_MLQ2620_MPS117_MC1100_L1p0", "ST", "TT", "DYJets", "QCD_Had", "WJets", "DATA"};

  TString histfolder_base = "Limits";

  vector<TString> systematics = {"nominal"};

  TString dir_base = PostAnalysisTool::m_finalselection_path;
  vector<TString> category_tags = {"cat1", "cat2"};
  vector<TString> histinname_base = {"stmet2", "stmet2"};
  vector<TString> histoutname_base = {"STMET", "STMET"};
  vector<TString> channel_tags = {"elch", "much", "elmuch", "tach"};


  vector<TString> syst_shift = {"up", "down"};
  vector<TString> syst_shift_combine = {"Up", "Down"};

  string command = "mkdir -p " + (string)m_datacard_path;
  system(command.c_str());

  TString outfilename = PostAnalysisTool::m_datacard_path + "combine_histograms.root";

  unique_ptr<TFile> outfile;
  outfile.reset(new TFile(outfilename, "RECREATE"));


  for(unsigned int category=0; category<category_tags.size(); category++){
    // cout << "category: " << category_tags[category] << endl;
    for(unsigned int channel=0; channel<channel_tags.size(); channel++){
      // cout << "channel: " << channel_tags[channel] << endl;

      for(unsigned int k=0; k<systematics.size(); k++){
        TString syst = systematics[k];
        cout << green << "  --> Syst: " << syst << reset << endl;

        for(unsigned int m=0; m<syst_shift.size(); m++){
          // cout << "syst_shift: " << syst_shift[m] << endl;
          TString dir_nom = dir_base;
          TString dir_scale = dir_base;
          TString dir_pdf = dir_base;

          if(syst == "nominal" && m > 0) break;
          dir_pdf += "pdf/";
          dir_scale += "scale/";

          for(unsigned int i=0; i<samples.size(); i++){
            // cout << "sample: " << samples[i] << endl;
            bool force_nominal = false;
            TString sample_in = samples[i];
            TString sample_out = samples[i];
            if(sample_out == "DATA") sample_out = "data_obs_" + PostAnalysisTool::m_year;
            cout << green << "    --> Sample: " << sample_in << reset << endl;

            if(sample_in == "DATA" && syst != "nominal") force_nominal = true;
            if(!(sample_in == "TT") && syst == "scale_TT") force_nominal = true;
            if(!(sample_in == "SingleTop") && syst == "scale_ST") force_nominal = true;
            if(!(sample_in == "DYJets") && syst == "scale_DYJets") force_nominal = true;
            if(!(sample_in == "WJets") && syst == "scale_WJets") force_nominal = true;

            TString type = "MC__";
            if(sample_in == "DATA") type = "DATA__";

            if(sample_in == "DATA"){
              if(channel_tags[channel] == "much" || channel_tags[channel] == "elmuch") sample_in += "_SingleMuon";
              else if(channel_tags[channel] == "elch") sample_in += "_SingleElectron";
              else sample_in += "_Tau";
            }

            TString filename = dir_nom + type + sample_in + ".root";
            if(!force_nominal){
              if(syst.Contains("scale_")) filename = dir_scale + sample_in + ".root";
              if(syst == "pdf") filename = dir_pdf + sample_in + ".root";
            }
            TFile* infile = new TFile(filename, "READ");

            TString histname_in = "";
            histname_in = channel_tags[channel] + "_" + category_tags[category] + "_" + syst + "_" + histfolder_base;
            if(force_nominal) histname_in = channel_tags[channel] + "_" + channel_tags[channel] + "_" + category_tags[category] + "_nominal" + "_" + histfolder_base;
            if(syst != "nominal" && !force_nominal) histname_in += "_" + syst_shift[m];
            histname_in += "/" + histinname_base[category];
            // think this is not needed, but not 100% sure
            // if(syst.Contains("scale") && !force_nominal) histname_in = channel_tags[channel] + "_" + histfolder_base + "_" + category_tags[category] + "_scale_" + syst_shift[m] +  "/" + histinname_base[category];

            TString histname_out =  histoutname_base[category] + "_" + channel_tags[channel] + "_" + category_tags[category] + "__" + sample_out;
            if(syst != "nominal") histname_out += "__" + syst + syst_shift_combine[m];

            // cout << "histname_in: " << histname_in << endl;
            TH1F* h_in = (TH1F*)infile->Get(histname_in);
            // if(sample_in.Contains("LQtoTMuM300") || sample_in.Contains("LQtoTMuM400")){
            //   h_in->Scale(1./10.);
            // }
            h_in->SetName(histname_out);
            outfile->cd();
            h_in->Write();
            infile->Close();
          }
        }
      }
    }
  }

  cout << green << "--> Finished producing the combine histograms in '" << outfilename << "'."  << reset << endl;



}

File: PostAnalyzer/functions/ProduceCombineHistograms.py

import os, sys, math, time
import subprocess
from utils import *
from collections import OrderedDict
from ROOT import TFile, TH1D

def ProduceCombineHistograms(self, signal_scaled_by=1.):
    print green('  --> Producing histograms for combine.')

    samples = self.signals + self.backgrounds + ['DATA']
    syst_shifts = [('up', 'Up'), ('down', 'Down')]

    ensureDirectory(self.combineinput_path)
    outfilename = os.path.join(self.combineinput_path, 'combine_histograms.root')
    outfile = TFile(outfilename, 'RECREATE')

    # loop
    for category in self.histnames_in_out_per_category:
        for channel in self.channels:
            for syst in self.systematics:
                print green('  --> Systematic: %s' % (syst))
                for shiftidx, syst_shift in enumerate(syst_shifts):
                    dir_nom = self.finalselection_path
                    dir_scale = os.path.join(self.finalselection_path, 'scale')
                    dir_pdf = os.path.join(self.finalselection_path, 'pdf')

                    if syst == 'nominal' and shiftidx > 0: continue
                    for sample in samples:
                        type = 'DATA__' if sample == 'DATA' else 'MC__'

                        force_nominal = False
                        sample_out = sample
                        if sample_out == 'DATA': sample_out = 'data_obs_' + self.year
                        else:                    sample_out += '_' + self.year
                        print green('    --> Sample: %s' % (sample))

                        if sample == 'DATA' and not syst == 'nominal': force_nominal = True
                        if not sample in self.processes_per_systematic.keys() and not syst == 'nominal': raise ValueError('When producing combine histograms: found systematic in systematic list that is not covered by \'processes_per_systematic\' dictionary. Please add it there and in the other dictionaries.')
                        if not syst == 'nominal' and not sample == self.processes_per_systematic[syst]: force_nominal = True
                        # if not sample == 'TT' and syst == 'scale_TT':  force_nominal = True
                        # if not sample == 'ST' and syst == 'scale_ST':  force_nominal = True
                        # if not sample == 'DYJets' and syst == 'scale_DYJets':  force_nominal = True
                        # if not sample == 'WJets' and syst == 'scale_WJets':    force_nominal = True

                        if sample == 'DATA': # choose correct PD
                            if channel == 'much' or channel == 'elmuch': sample += '_SingleMuon'
                            elif channel == 'elch': sample += '_SingleElectron'
                            else: sample += '_Tau'

                        infilename = os.path.join(dir_nom, type+sample+'.root')
                        if not force_nominal:
                            if 'scale_' in syst: infilename = os.path.join(dir_scale, sample_in+'.root')
                            if syst == 'pdf':    infilename = os.path.join(dir_pdf, sample_in+'.root')
                        infile = TFile(infilename, 'READ')

                        histname_in = '_'.join([channel, category, syst, self.limithisttag])
                        if force_nominal: histname_in = '_'.join([channel, category, 'nominal', self.limithisttag])
                        if not syst == 'nominal' and not force_nominal: histname_in = '_'.join([histname_in, syst_shift[0]])
                        histname_in += '/' + self.histnames_in_out_per_category[category][0]
                        # think this is not needed, but not 100% sure
                        # if(syst.Contains("scale") && !force_nominal) histname_in = self.channels[channel] + "_" + histfolder_base + "_" + category_tags[category] + "_scale_" + syst_shift[m] +  "/" + histinname_base[category];

                        histname_out = '_'.join([self.histnames_in_out_per_category[category][1], channel, category]) + '__' + sample_out
                        if not syst == 'nominal': histname_out += '__' + syst + syst_shift[1]
                        h_in = infile.Get(histname_in)
                        h_in.SetName(histname_out)
                        h_in.Scale(signal_scaled_by)
                        outfile.cd()
                        h_in.Write()
                        infile.Close()

    print green('  --> Finished producing the combine histograms in \'%s\'.' % (outfilename))
